### PR TITLE
Allow array spread and rest params

### PR DESCRIPTION
**Pull request by rcebulko**

### CODE REVIEW

**Comment by rcebulko** [comment]
@jridgewell bumping this in your GH queue

File: src/core/assert/base.js [hunk 1]
```diff
@@ -57,18 +57,13 @@ export function assert(
...
-  let i = 3;
-
   // Substitute provided values into format string in message
   const splitMessage = opt_message.split('%s');
   let message = splitMessage.shift();
   const messageArray = [message];
 
   while (splitMessage.length) {
-    const subValue = arguments[i++];
+    const subValue = msgArgs.shift();
```
Comment (jridgewell) [inline src/core/assert/base.js:66]: This turns an O(n) into O(n<sup>2</sup>) operation.

File: src/core/assert/base.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ export function assert(
   sentinel,
   shouldBeTruthy,
   opt_message = 'Assertion failed',
-  var_args
+  ...msgArgs
```
Comment (jridgewell) [inline src/core/assert/base.js:49]: This is like considerably slower. `…args` should never be used unless you explicitly need access to the array methods, or need to pass the arguments outside of the current invocation.

File: src/core/error.js [hunk 3]
```diff
@@ -37,15 +37,14 @@ export function duplicateErrorIfNecessary(error) {
...
  * @return {!Error}
  * @visibleForTesting
  */
-export function createErrorVargs(var_args) {
+export function createErrorVargs(...args) {
   let error = null;
   let message = '';
-  for (let i = 0; i < arguments.length; i++) {
-    const arg = arguments[i];
+  for (const arg of args) {
```
Comment (jridgewell) [inline src/core/error.js:47]: Just iterate `arguments`.

File: src/core/error.js [hunk 4]
```diff
@@ -67,10 +66,10 @@ export function createErrorVargs(var_args) {
 /**
  * Rethrows the error without terminating the current context. This preserves
  * whether the original error designation is a user error or a dev error.
- * @param {...*} var_args
+ * @param {...*} args
  */
-export function rethrowAsync(var_args) {
-  const error = createErrorVargs.apply(null, arguments);
+export function rethrowAsync(...args) {
+  const error = createErrorVargs(...args);
```
Comment (jridgewell) [inline src/core/error.js:72]: Doing `...args` to collect args into an array, then `...args` to spread it, is considerably slower than just `...arguments`. Prefer `arguments` whenever possible.

File: src/core/types/function.js [hunk 5]
```diff
@@ -36,7 +36,7 @@ export function once(fn) {
   let callback = fn;
   return (...args) => {
```
Comment (jridgewell) [inline src/core/types/function.js:37]: Ugh, this shouldn't have been here to start with.

File: src/core/types/function.js [hunk 6]
```diff
@@ -36,7 +36,7 @@ export function once(fn) {
...
     if (!evaluated) {
-      retValue = callback.apply(self, args);
+      retValue = callback(...args);
```
Comment (jridgewell) [inline src/core/types/function.js:39]: Ditto the other `arguments` comments.

**Review by jridgewell — CHANGES_REQUESTED** [review]
Needs updates to handle `ArrayPattern`:
- build-system/babel-plugins/babel-plugin-transform-default-assignment/index.js
- build-system/eslint-rules/no-deep-destructuring.js
- build-system/eslint-rules/prefer-destructuring.js
- build-system/eslint-rules/unused-private-field.js

Can be removed:
- build-system/babel-plugins/babel-plugin-transform-simple-array-destructure/index.js

File: src/core/assert/base.js (hunk 1 above)
Comment (rcebulko) [reply src/core/assert/base.js]: Reverting all changes to this file

File: src/core/error.js (hunk 3 above)
Comment (rcebulko) [reply src/core/error.js]: Done, removed spread changes in this file.

File: src/core/error.js (hunk 4 above)
Comment (rcebulko) [reply src/core/error.js]: Reverted;I mistakenly assumed that the implementation would be more favorable to such a standard use of the feature 🤦🏻

File: src/core/types/function.js (hunk 5 above)
Comment (rcebulko) [reply src/core/types/function.js]: Yeah I'd been wondering about this, and if it caused the iterator polyfill to be included regardless. Should I rewrite to use `arguments`?

File: src/core/types/function.js (hunk 6 above)
Comment (rcebulko) [reply src/core/types/function.js]: ACK

**Comment by rcebulko** [comment]
> Needs updates to handle `ArrayPattern`:
> 
> * build-system/babel-plugins/babel-plugin-transform-default-assignment/index.js
> * build-system/eslint-rules/no-deep-destructuring.js
> * build-system/eslint-rules/prefer-destructuring.js
> * build-system/eslint-rules/unused-private-field.js
> 
> Can be removed:
> 
> * build-system/babel-plugins/babel-plugin-transform-simple-array-destructure/index.js

I can't make heads or tails of a lot of this code. @samouri do you have insights here? Would be happy to pair-program but absolutely lost at first glance

**Comment by jridgewell** [comment]
I can help later this week. Changes LGTM.

**Pull request closed without merging**